sqlx-core/src/type_checking.rs

```diff
@@ -0,0 +1,188 @@
+use crate::database::Database;
+use crate::decode::Decode;
+use crate::type_info::TypeInfo;
+use crate::value::Value;
+use std::any::Any;
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+
+/// The type of query parameter checking done by a SQL database.
+#[derive(PartialEq, Eq)]
+pub enum ParamChecking {
+    /// Parameter checking is weak or nonexistent (uses coercion or allows mismatches).
+    Weak,
+    /// Parameter checking is strong (types must match exactly).
+    Strong,
+}
+
+/// Type-checking extensions for the `Database` trait.
+///
+/// Mostly supporting code for the macros, and for `Debug` impls.
+pub trait TypeChecking: Database {
+    /// Describes how the database in question typechecks query parameters.
+    const PARAM_CHECKING: ParamChecking;
+
+    /// Get the full path of the Rust type that corresponds to the given `TypeInfo`, if applicable.
+    ///
+    /// If the type has a borrowed equivalent suitable for query parameters,
+    /// this is that borrowed type.
+    fn param_type_for_id(id: &Self::TypeInfo) -> Option<&'static str>;
+
+    /// Get the full path of the Rust type that corresponds to the given `TypeInfo`, if applicable.
+    ///
+    /// Always returns the owned version of the type, suitable for decoding from `Row`.
+    fn return_type_for_id(id: &Self::TypeInfo) -> Option<&'static str>;
+
+    /// Get the name of the Cargo feature gate that must be enabled to process the given `TypeInfo`,
+    /// if applicable.
+    fn get_feature_gate(info: &Self::TypeInfo) -> Option<&'static str>;
+
+    /// If `value` is a well-known type, decode and format it using `Debug`.
+    ///
+    /// If `value` is not a well-known type or could not be decoded, the reason is printed instead.
+    fn fmt_value_debug(value: &<Self as Database>::Value) -> FmtValue<'_, Self>;
```

Comment:
Can the reference to `Value` be replaced with `ValueRef`?

Reply:
The reason why it doesn't is explained at line 59.

It's feasible to add a `Clone` impl to `SqliteValueRef` but I would almost prefer to add like a `reborrow()` method to `ValueRef` instead, but either of those changes were rather out of scope for this PR.

Since it turned out to be a breaking change anyway, I'd be inclined to revisit that.